### ISSUE TITLE
ARM(32) Walker should not use FP unwinding (should not rely on valid LR)

### ISSUE THREAD

**Issue by Swatinem** [luser/rust-minidump#460]
Context:

We have seen some minidumps in the wild which have an obvious addr to something on the stack stored in the LR.

In `get_caller_by_frame_pointer`, the ARM stackwalker just blindly uses the LR of the callee without validation:
https://github.com/luser/rust-minidump/blob/7f5adf743560705fa3d46a784c874655e6681432/minidump-processor/src/stackwalker/arm.rs#L140

Though the LR is not guarenteed to be valid. At least the ARMv6 and ARMv7 architecture manuals say that LR can be used as a general purpose register:

![grafik](https://user-images.githubusercontent.com/580492/154690152-15f77c85-ad2b-4f64-9380-94f5c1ebbe50.png)

(thats from the ARMv7 manual)

Breakpad special cases this in a weird way:

https://github.com/getsentry/breakpad/blob/63c90f00bf64a1b4a8883e50bbe8217048ed2a85/src/processor/stackwalker.cc#L300-L302

This is a bit of a misnomer. Essentially making the `fp_register` invalid completely disabled FP unwinding on ARM on anything other than iOS, probably because the iOS calling conventions are stricter than the ARM conventions.

**TLDR**:

We should either not do FP unwinding on (non-iOS) ARM at all, or more strictly validate the LR value (it should pass the "address looks valid" check).

**Comment by Gankra** [comment]
My comment there could be written a bit better but yeah you want to trust frame pointers to put you into "unknown" code at a good place for scanning to be more likely to work. This ideally includes:

* modules we don't have cfi for ✔️ 
* modules that have been unloaded(?) ❌ 
* JIT pages (not mapped to any module) ❌ 

Currently instruction_seems_valid only accepts the first one. That said, because scanning only works when instruction_seems_valid, allowing the extra two cases only works if we dip into them for exactly one frame. Otherwise there's no(?) advantage to just always requiring instruction_seems_valid.

The other issue is that we want to be able to have cfi or fp techniques report "the unwind is complete". Unfortunately this is not like, a first-class concept so we generally just infer it from stuff like "return address is roughly null". These are addresses instruction_seems_valid would reject, so at very least you need to carefully try to identify and propagate these "sentinel" results.

Unfortunately those sentinel results are also sometimes wrong:

https://bugzilla.mozilla.org/show_bug.cgi?id=1747821

This is all to say that I find these heuristics *deeply* frustrating and wobbly, and wish we had like 10x more tests of interesting situations we were confident in the desired result for 😭

**Comment by mitsuhiko** [comment]
> Breakpad special cases this in a weird way:

**Comment by luser** [comment]
Firefox needs some sort of "Donate my crash reports to science" opt-in setting in about:config that grants Mozilla permission to use sanitized copies of minidump files from that user as test data. :)

**Comment by Swatinem** [comment]
Hm, maybe I didn’t really express myself in an understandable way.

This is not really about frame pointer unwinding, and about frame pointers at all.

It is rather that the "frame pointer unwinding" heuristic just says "return addr = LR", which is incorrect since on all ARM platforms but apple, LR is a general purpose register (with a special role) and the compiler is free to use it inside of functions however it sees fit.

@loewenheim can give you the example that we have (context frame + bogus recovered frame via "fp" method)

And again, I see two solutions to this in order to solve the false positives:

* Do not use the "fp heuristic" *at all* for all ARM platforms except apple (which is more strict about LR not being available as general purpose register) this is the same that breakpad does
* OR: sanity-check the LR register value before using it (for example, checking that it points either *to* a mapped module; or at the very least **not into** the stack memory). This can have mixed results, as it is still a bit prone to false positives, but might as well yield better results in cases the compiler did not use the LR register as general purpose.

**Comment by gabrielesvelto** [comment]
> Firefox needs some sort of "Donate my crash reports to science" opt-in setting in about:config that grants Mozilla permission to use sanitized copies of minidump files from that user as test data. :)

I wonder if we could use crashes happening in automation; those do not contain PII by definition.

**Comment by loewenheim** [comment]
Here's an example of a stacktrace exhibiting the problem @Swatinem is describing. This is from a customer minidump.

Registers are the same between breakpad and rust-minidump:
```
{
      "fp": "0xa5cd671c",
      "lr": "0xa5cd70a0",
      "pc": "0x420cc04c",
      "r0": "0x1",
      "r1": "0x0",
      "r10": "0xa6f1c4c0",
      "r12": "0x0",
      "r2": "0xa5cd6598",
      "r3": "0x0",
      "r4": "0x1",
      "r5": "0x0",
      "r6": "0xa5cd6598",
      "r7": "0x109",
      "r8": "0x1",
      "r9": "0x0",
      "sp": "0xa5cd6580"
}
```
Stacktrace returned by breakpad:
```
[
      {
        "instruction_addr": "0x420cc04c",
        "package": "libc.so.6",
        "trust": "context"
      },
      {
        "instruction_addr": "0xa3fcc0",
        "package": "/navigation/opt/irobot/bin/cleantrack",
        "trust": "scan"
      },
      {
        "instruction_addr": "0x9f4b48",
        "package": "/navigation/opt/irobot/bin/cleantrack",
        "trust": "scan"
      },
      {
        "instruction_addr": "0x989680",
        "package": "/navigation/opt/irobot/bin/cleantrack",
        "trust": "scan"
      },
      {
        "instruction_addr": "0x1e70430",
        "package": "/navigation/opt/irobot/bin/cleantrack",
        "trust": "scan"
      },
      {
        "instruction_addr": "0x1f60328",
        "package": "/navigation/opt/irobot/bin/cleantrack",
        "trust": "scan"
      },
      {
        "instruction_addr": "0x424e27ac",
        "package": "libstdc++.so.6",
        "trust": "scan"
      },
      {
        "instruction_addr": "0x1f60040",
        "package": "/navigation/opt/irobot/bin/cleantrack",
        "trust": "scan"
      },
      {
        "instruction_addr": "0x424515b8",
        "package": "libstdc++.so.6",
        "trust": "scan"
      },
      {
        "instruction_addr": "0x421661d0",
        "package": "libpthread.so.0",
        "trust": "scan"
      },
      {
        "instruction_addr": "0x4207486f",
        "package": "libc.so.6",
        "trust": "scan"
      },
      {
        "instruction_addr": "0xa5dc4c27",
        "package": "libfastrtps.so.2.3",
        "trust": "scan"
      }
]
```
Stacktrace returned by rust-minidump:
```
[
      {
        "instruction_addr": "0x420cc04c",
        "package": "libc.so.6",
        "trust": "context"
      },
      {
        "instruction_addr": "0xa5cd70a0",
        "trust": "fp"
      }
]
```
